### PR TITLE
API-373+376: Partial update one or a list of variant product 

### PR DESCRIPTION
**Pull request by solivier**
**Description (for Contributor and Core Developer)**

This PR adds tests for a partial update of one and a list of variant products.
It also fixes some bugs in variant products:
- common attributes should be immutable.
- Fix an order bug (when family comes before the Parent, the parent erase the family so the validation is ok but should not)
- Fix bug when parent is set to nul.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Yes
| Added Behats                      | No
| Added integration tests           | Yes
| Changelog updated                 | Yes
| Review and 2 GTM                  | No
| Micro Demo to the PO (Story only) | Yes
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: CHANGELOG-2.0.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# 2.0.x
+# 2.0.1 (2017-10-05)
```
Comment (willy-ahva) [inline CHANGELOG-2.0.md]: nope

File: composer.json [hunk 2]
```diff
@@ -147,7 +147,8 @@
             }
         },
         "branch-alias": {
-            "dev-master": "2.0.x-dev",
+            "dev-master": "2.1.x-dev",
+            "dev-2.0": "2.0.x-dev",
```
Comment (willy-ahva) [inline composer.json]: nope nope

File: src/Pim/Bundle/ApiBundle/Controller/ProductController.php [hunk 3]
```diff
@@ -748,4 +761,32 @@ protected function searchAfterIdentifier(
 
         return $paginatedProducts;
     }
+
+    /**
+     * This method order the data by setting the parent field first. It comes from the ParentFieldSetter that sets the
+     * family from the parent if the product family is null. By doing this the validator does not fail if the family
+     * field has been set to nul from the API. So to prevent this we order the parent before the family field. this way
```
Comment (willy-ahva) [inline src/Pim/Bundle/ApiBundle/Controller/ProductController.php]: has been set to null*

File: src/Pim/Bundle/CatalogBundle/Version.php [hunk 4]
```diff
@@ -12,7 +12,7 @@
 class Version
 {
     /** @staticvar string */
-    const VERSION = '2.0.0';
+    const VERSION = '2.0.1';
```
Comment (willy-ahva) [inline src/Pim/Bundle/CatalogBundle/Version.php]: Did you miss a rebase or something like that? I don't understand why you have these diff

File: src/Pim/Component/Catalog/ProductModel/Filter/AttributeFilterInterface.php [hunk 5]
```diff
@@ -1,7 +1,7 @@
 <?php
 declare(strict_types=1);
 
-namespace Pim\Component\Connector\Processor\Denormalization\AttributeFilter;
+namespace Pim\Component\Catalog\ProductModel\Filter;
```
Comment (willy-ahva) [inline src/Pim/Component/Catalog/ProductModel/Filter/AttributeFilterInterface.php]: Why in ProductModel ? It `Filters imported data from products, variant products and product models.`. Product model namespace seems not be accurate in this case.

File: src/Pim/Component/Catalog/ProductModel/Filter/ProductAttributeFilter.php [hunk 6]
```diff
@@ -1,11 +1,13 @@
 <?php
 declare(strict_types=1);
 
-namespace Pim\Component\Connector\Processor\Denormalization\AttributeFilter;
+namespace Pim\Component\Catalog\ProductModel\Filter;
```
Comment (willy-ahva) [inline src/Pim/Component/Catalog/ProductModel/Filter/ProductAttributeFilter.php]: Same comment

File: src/Pim/Component/Catalog/ProductModel/Filter/AttributeFilterInterface.php (hunk 5 above)
Comment (willy-ahva) [reply src/Pim/Component/Catalog/ProductModel/Filter/AttributeFilterInterface.php]: +CHANGELOG

File: src/Pim/Component/Catalog/ProductModel/Filter/ProductAttributeFilter.php (hunk 6 above)
Comment (willy-ahva) [reply src/Pim/Component/Catalog/ProductModel/Filter/ProductAttributeFilter.php]: +CHANGELOG

File: src/Pim/Component/Catalog/ProductModel/Filter/ProductAttributeFilter.php [hunk 7]
```diff
@@ -28,43 +30,58 @@ class ProductAttributeFilter implements AttributeFilterInterface
...
      */
     public function __construct(
         IdentifiableObjectRepositoryInterface $productModelRepository,
-        IdentifiableObjectRepositoryInterface $familyRepository
+        IdentifiableObjectRepositoryInterface $familyRepository,
+        IdentifiableObjectRepositoryInterface $productRepository
     ) {
         $this->productModelRepository = $productModelRepository;
         $this->familyRepository = $familyRepository;
+        $this->productRepository = $productRepository;
```
Comment (willy-ahva) [inline src/Pim/Component/Catalog/ProductModel/Filter/ProductAttributeFilter.php]: +CHANGELOG

File: src/Pim/Component/Catalog/Validator/Constraints/ImmutableVariantAxesValues.php [hunk 8]
```diff
@@ -17,6 +17,9 @@ class ImmutableVariantAxesValues extends Constraint
 {
     public const UPDATED_VARIANT_AXIS_VALUE = 'pim_catalog.constraint.modified_variant_axis_value';
 
+    /** @var string */
+    public $propertyPath = 'attribute';
```
Comment (willy-ahva) [inline src/Pim/Component/Catalog/Validator/Constraints/ImmutableVariantAxesValues.php]: Shouldn't it be protected ?

File: src/Pim/Component/Connector/Processor/Denormalization/ProductModelProcessor.php [hunk 9]
```diff
@@ -56,14 +56,14 @@ class ProductModelProcessor extends AbstractProcessor implements ItemProcessorIn
...
-     * @param ObjectDetacherInterface               $objectDetacher
-     * @param AttributeFilterInterface              $productModelAttributeFilter
-     * @param string                                $importType
+     * @param SimpleFactoryInterface                                              $productModelFactory
+     * @param ObjectUpdaterInterface                                              $productModelUpdater
+     * @param IdentifiableObjectRepositoryInterface                               $productModelRepository
+     * @param ValidatorInterface                                                  $validator
+     * @param FilterInterface                                                     $productModelFilter
+     * @param ObjectDetacherInterface                                             $objectDetacher
+     * @param \Pim\Component\Catalog\ProductModel\Filter\AttributeFilterInterface $productModelAttributeFilter
```
Comment (willy-ahva) [inline src/Pim/Component/Connector/Processor/Denormalization/ProductModelProcessor.php]: Add in using statements

File: src/Pim/Component/Connector/Processor/Denormalization/ProductProcessor.php [hunk 10]
```diff
@@ -53,14 +54,14 @@ class ProductProcessor extends AbstractProcessor implements ItemProcessorInterfa
...
-     * @param FilterInterface                       $productFilter
-     * @param AttributeFilterInterface              $productAttributeFilter
+     * @param IdentifiableObjectRepositoryInterface                               $repository
+     * @param ProductBuilderInterface                                             $productBuilder
+     * @param ProductBuilderInterface                                             $variantProductBuilder
+     * @param ObjectUpdaterInterface                                              $updater
+     * @param ValidatorInterface                                                  $validator
+     * @param ObjectDetacherInterface                                             $detacher
+     * @param FilterInterface                                                     $productFilter
+     * @param \Pim\Component\Catalog\ProductModel\Filter\AttributeFilterInterface $productAttributeFilter
```
Comment (willy-ahva) [inline src/Pim/Component/Connector/Processor/Denormalization/ProductProcessor.php]: Add in using statements

File: src/Pim/Component/Catalog/ProductModel/Filter/AttributeFilterInterface.php (hunk 5 above)
Comment (solivier) [reply src/Pim/Component/Catalog/ProductModel/Filter/AttributeFilterInterface.php]: Well it's mainly to merge the data from the product model into the product variant.

File: src/Pim/Component/Connector/Processor/Denormalization/ProductModelProcessor.php (hunk 9 above)
Comment (solivier) [reply src/Pim/Component/Connector/Processor/Denormalization/ProductModelProcessor.php]: nope we should access it from another classes

File: src/Pim/Component/Catalog/Validator/Constraints/ImmutableVariantAxesValues.php (hunk 8 above)
Comment (solivier) [reply src/Pim/Component/Catalog/Validator/Constraints/ImmutableVariantAxesValues.php]: nope we should access it from another classes

File: src/Pim/Component/Catalog/ProductModel/Filter/ProductAttributeFilter.php (hunk 7 above)
Comment (solivier) [reply src/Pim/Component/Catalog/ProductModel/Filter/ProductAttributeFilter.php]: Done

File: src/Pim/Component/Catalog/ProductModel/Filter/ProductAttributeFilter.php (hunk 6 above)
Comment (solivier) [reply src/Pim/Component/Catalog/ProductModel/Filter/ProductAttributeFilter.php]: Done

File: src/Pim/Component/Catalog/ProductModel/Filter/ProductAttributeFilter.php [hunk 11]
```diff
@@ -28,43 +29,58 @@ class ProductAttributeFilter implements AttributeFilterInterface
...
     /**
      * {@inheritdoc}
      */
-    public function filter(array $flatProduct): array
+    public function filter(array $standardProduct): array
     {
-        if (isset($flatProduct['parent']) &&
-            null !== $parentProductModel = $this->productModelRepository->findOneByIdentifier($flatProduct['parent'])
+        if (!array_key_exists('identifier', $standardProduct)) {
+            throw new \InvalidArgumentException('Missing identifier key');
```
Comment (skeleton) [inline src/Pim/Component/Catalog/ProductModel/Filter/ProductAttributeFilter.php]: find (or create if it does not exist) a business exception

File: src/Pim/Bundle/ApiBundle/Controller/ProductController.php [hunk 12]
```diff
@@ -315,7 +317,12 @@ public function partialUpdateAction(Request $request, $code): Response
 
         if ($isCreation) {
             $this->validateCodeConsistency($code, $data);
-            $product = $this->productBuilder->createProduct($code);
+
+            if (isset($data['parent'])) {
+                $product = $this->variantProductBuilder->createProduct($data['identifier']);
+            } else {
+                $product = $this->productBuilder->createProduct();
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Controller/ProductController.php]: same

File: src/Pim/Bundle/ApiBundle/Controller/ProductController.php [hunk 13]
```diff
@@ -315,7 +317,12 @@ public function partialUpdateAction(Request $request, $code): Response
 
         if ($isCreation) {
             $this->validateCodeConsistency($code, $data);
-            $product = $this->productBuilder->createProduct($code);
+
+            if (isset($data['parent'])) {
+                $product = $this->variantProductBuilder->createProduct($data['identifier']);
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Controller/ProductController.php]: createProduct($code) (as it's a patch, `identifier` key is not mandatory on body)

File: src/Pim/Component/Catalog/ProductModel/Filter/ProductAttributeFilter.php (hunk 11 above)
Comment (skeleton) [reply src/Pim/Component/Catalog/ProductModel/Filter/ProductAttributeFilter.php]: `Missing "identifier" key`

**Pull request merged**